### PR TITLE
Models API with Inference Routing

### PR DESCRIPTION
**Pull request by yanxi0830**
# Changes
✅  add models api endpoints for querying models distribution
✅  add special router "models-router" for inference endpoints to dynamically route to different models 
✅  configure script to setup models_config for models endpoint, and infer model config
✅  current single adapter based implementation is still in tact

-- not in this PR
🚧  register_model endpoint to add new models
🚧  safety router to run prompt_guard / llama_guard
🚧  client CLI for querying distribution models endpoint --> will be in llama-stack-apps

# run.yaml
```yaml
built_at: '2024-09-18T13:41:17.656743'
image_name: local
docker_image: null
conda_env: local
apis_to_serve:
- inference
- memory
- safety
- telemetry
- agents
- models
provider_map:
  inference: models-router
  safety:
    provider_id: meta-reference
    config:
      llama_guard_shield:
        model: Llama-Guard-3-8B
        excluded_categories: []
        disable_input_check: false
        disable_output_check: false
      prompt_guard_shield:
        model: Prompt-Guard-86M
  telemetry:
    provider_id: meta-reference
    config: {}
  agents:
    provider_id: meta-reference
    config: {}
  memory:
    provider_id: meta-reference
    config: {}
  models:
    provider_id: builtin
    config:
      models_config:
      - core_model_id: Meta-Llama3.1-8B-Instruct
        provider_id: meta-reference
        api: inference
        config:
          model: Meta-Llama3.1-8B-Instruct
          quantization: null
          torch_seed: null
          max_seq_len: 4096
          max_batch_size: 1
      - core_model_id: Meta-Llama3.1-8B
        provider_id: meta-reference
        api: inference
        config:
          model: Meta-Llama3.1-8B
          quantization: null
          torch_seed: null
          max_seq_len: 4096
          max_batch_size: 1
      - core_model_id: Llama-Guard-3-8B
        provider_id: meta-reference
        api: safety
        config:
          model: Llama-Guard-3-8B
          excluded_categories: []
          disable_input_check: false
          disable_output_check: false
      - core_model_id: Prompt-Guard-86M
        provider_id: meta-reference
        api: safety
        config:
          model: Prompt-Guard-86M
```
# Test w/ Client
```
python -m llama_stack.apis.inference.client
```
- Server: local inference with 2 different models

image

image

```
python -m llama_stack.apis.models.clients
```
image
image

# Test w/ SDK
```
python sdk_examples/agents/client.py
```
- agents still works w/ routing

# Test regular simple example support
```
llama stack build
llama stack configure simple-local
```
- configure will generate model_config specs based on inference/safety configurations.

### CODE REVIEW

File: llama_stack/configs/examples/local-router-run.yaml [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+      prompt_guard_shield:
+        model: Prompt-Guard-86M
+  telemetry:
+    provider_id: meta-reference
+    config: {}
+  agents:
+    provider_id: meta-reference
+    config: {}
+  models:
+    provider_id: builtin
```
Comment (ashwinb) [inline llama_stack/configs/examples/local-router-run.yaml:31]: initial impression: this feels odd and should not be needed to be seen by the user. 

also config -> models_config should just be config

File: llama_stack/apis/models/models.py [hunk 2]
```diff
@@ -1,14 +1,49 @@
...
 
-from pydantic import BaseModel  # noqa: F401
+from llama_models.schema_utils import json_schema_type, webmethod
+from llama_stack.distribution.datatypes import GenericProviderConfig
+from pydantic import BaseModel, Field
 
 
-class Models(Protocol): ...
+@json_schema_type
+class ModelSpec(BaseModel):
```
Comment (ashwinb) [inline llama_stack/apis/models/models.py:14]: options for name:
- ServableModel
- ServedModel
- ModelServingSpec

File: llama_stack/apis/models/models.py [hunk 3]
```diff
@@ -1,14 +1,49 @@
...
+    llama_model_metadata: Model = Field(
```
Comment (ashwinb) [inline llama_stack/apis/models/models.py:15]: just `llama_model` works

File: llama_stack/apis/models/models.py [hunk 4]
```diff
@@ -1,14 +1,49 @@
...
+        description="All metadatas associated with llama model (defined in llama_models.models.sku_list). "
+    )
+    provider_id: str = Field(
+        description="API provider that is serving this model (e.g. meta-reference, local)",
+    )
+    api: str = Field(
+        description="API that this model is serving (e.g. inference / safety)",
```
Comment (ashwinb) [inline llama_stack/apis/models/models.py:22]: `default="inference"`?

File: llama_stack/apis/models/models.py [hunk 5]
```diff
@@ -1,14 +1,49 @@
...
+from pydantic import BaseModel, Field
 
 
-class Models(Protocol): ...
+@json_schema_type
+class ModelSpec(BaseModel):
+    llama_model_metadata: Model = Field(
+        description="All metadatas associated with llama model (defined in llama_models.models.sku_list). "
+    )
+    provider_id: str = Field(
```
Comment (ashwinb) [inline llama_stack/apis/models/models.py:18]: could inherit from `GenericProviderConfig`

File: llama_stack/configs/examples/local-router-run.yaml [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+docker_image: null
+conda_env: local
+apis_to_serve:
+- inference
+- memory
+- safety
+- telemetry
+- agents
+- models
+provider_map:
```
Comment (yanxi0830) [inline llama_stack/configs/examples/local-router-run.yaml:12]:
```
routing_table:
  inference: 
    - core_model_id: Meta-Llama3.1-8B-Instruct
        provider_id: meta-reference
        api: inference
        config:
          model: Meta-Llama3.1-8B-Instruct
          quantization: null
          torch_seed: null
          max_seq_len: 4096
          max_batch_size: 1
  safety --> shield: 
    - provider: 
  
```

**Comment by yanxi0830** [comment]
Continued in https://github.com/meta-llama/llama-stack/pull/85

**Pull request closed without merging**